src/lib/comms/sol-connman-impl-connman.c

```diff
@@ -0,0 +1,910 @@
+    SOL_INT_CHECK(r, < 0, r);
+
+    r = sd_bus_message_read_basic(m, SD_BUS_TYPE_STRING, &state);
+    SOL_INT_CHECK(r, < 0, r);
+
+    if (state) {
+        if (!strcmp(state, "online"))
+            _ctx.connman_state = SOL_CONNMAN_STATE_ONLINE;
+        else if (!strcmp(state, "ready"))
+            _ctx.connman_state = SOL_CONNMAN_STATE_READY;
+        else if (!strcmp(state, "idle"))
+            _ctx.connman_state = SOL_CONNMAN_STATE_IDLE;
+        else if (!strcmp(state, "offline"))
+            _ctx.connman_state = SOL_CONNMAN_STATE_OFFLINE;
+        else
+            _ctx.connman_state = SOL_CONNMAN_STATE_UNKNOWN;
+    }
+
+    r = sd_bus_message_exit_container(m);
+    SOL_INT_CHECK(r, < 0, r);
+
+    return 0;
+}
+
+static int
+services_list_changed(sd_bus_message *m)
+{
+    int r = 0;
+    char *path = NULL;
+
+    r = sd_bus_message_enter_container(m, SD_BUS_TYPE_ARRAY, "(oa{sv})");
+    SOL_INT_CHECK(r, < 0, r);
+
+    while (sd_bus_message_enter_container(m, SD_BUS_TYPE_STRUCT, "oa{sv}") > 0) {
+        r = sd_bus_message_read(m, "o", &path);
+        SOL_INT_CHECK(r, < 0, r);
+
+        r = get_services_properties(m, path);
+        SOL_INT_CHECK(r, < 0, r);
+
+        r = sd_bus_message_exit_container(m);
+        SOL_INT_CHECK(r, < 0, r);
+    }
+
+    r = sd_bus_message_exit_container(m);
+    SOL_INT_CHECK(r, < 0, r);
+
+    r = sd_bus_message_enter_container(m, SD_BUS_TYPE_ARRAY, "o");
+    SOL_INT_CHECK(r, < 0, r);
+
+    while (sd_bus_message_read(m, "o", &path) > 0) {
+        remove_services(path);
+    }
+
+    r = sd_bus_message_exit_container(m);
+    SOL_INT_CHECK(r, < 0, r);
+
+    return 0;
+}
+
+static int
+_services_properties_changed(sd_bus_message *m, void *userdata,
+    sd_bus_error *ret_error)
+{
+    struct ctx *pending = userdata;
+
+    pending->service_slot = sd_bus_slot_unref(pending->service_slot);
+
+    if (sol_bus_log_callback(m, userdata, ret_error) < 0)
+        return -EINVAL;
+
+    return services_list_changed(m);
+}
+
+
+
+static int
+_manager_properties_changed(sd_bus_message *m, void *userdata,
+    sd_bus_error *ret_error)
+{
+    struct ctx *pending = userdata;
+    char *str;
+    int r;
+
+    pending->manager_slot = sd_bus_slot_unref(pending->manager_slot);
+
+    if (sol_bus_log_callback(m, userdata, ret_error) < 0)
+        return -EINVAL;
+
+    r = sd_bus_message_enter_container(m, SD_BUS_TYPE_ARRAY, "{sv}");
+    SOL_INT_CHECK(r, < 0, r);
+
+    do {
+
+        r = sd_bus_message_enter_container(m, SD_BUS_TYPE_DICT_ENTRY, "sv");
+        SOL_INT_CHECK_GOTO(r, < 1, end);
+
+        r = sd_bus_message_read_basic(m, SD_BUS_TYPE_STRING, &str);
+        SOL_INT_CHECK(r, < 0, r);
+
+        if (strcmp(str, "State") == 0) {
+            r = get_manager_properties(m);
+            SOL_INT_CHECK(r, < 0, r);
+        } else {
+            r = sd_bus_message_skip(m, "v");
+            SOL_INT_CHECK(r, < 0, r);
+        }
+
```

Comment:
as said in the .h, here you would inform the generic monitor to say something changed (it could be state or offline.

also log what you are skipping `SOL_DBG()` and AFAIU you should get the OfflineMode.